### PR TITLE
Add constants and helpers for ES Internal Users

### PR DESCRIPTION
**Pull request by raihankhan**
Signed-off-by: raihankhan <raihan@appscode.com>

### CODE REVIEW

File: apis/kubedb/v1alpha2/constants.go [hunk 1]
```diff
@@ -347,18 +347,21 @@ const (
...
+	AllReplicasAreReady                        = "AllReplicasReady"
+	SomeReplicasAreNotReady                    = "SomeReplicasNotReady"
+	DatabaseAcceptingConnectionRequest         = "DatabaseAcceptingConnectionRequest"
+	DatabaseNotAcceptingConnectionRequest      = "DatabaseNotAcceptingConnectionRequest"
+	ReadinessCheckSucceeded                    = "ReadinessCheckSucceeded"
+	ReadinessCheckFailed                       = "ReadinessCheckFailed"
+	DatabaseProvisioningStartedSuccessfully    = "DatabaseProvisioningStartedSuccessfully"
+	DatabaseSuccessfullyProvisioned            = "DatabaseSuccessfullyProvisioned"
+	DatabaseHaltedSuccessfully                 = "DatabaseHaltedSuccessfully"
+	InternalUsersReady                         = "InternalUsersReady"
```
Comment (kamolhasan) [inline apis/kubedb/v1alpha2/constants.go:362]: Type : `InternalUsersReady` --> `InternalUsersSynced`

File: apis/kubedb/v1alpha2/elasticsearch_helpers.go [hunk 2]
```diff
@@ -595,6 +595,70 @@ func (e *Elasticsearch) setDefaultInternalUsersAndRoleMappings(esVersion *catalo
 		return
 	}
 
+	if esVersion.Spec.AuthPlugin == catalog.ElasticsearchAuthPluginXpack {
```
Comment (kamolhasan) [inline apis/kubedb/v1alpha2/elasticsearch_helpers.go:598]: Also add the version check

File: apis/kubedb/v1alpha2/elasticsearch_helpers.go [hunk 3]
```diff
@@ -595,6 +595,70 @@ func (e *Elasticsearch) setDefaultInternalUsersAndRoleMappings(esVersion *catalo
...
+		inUsers := e.Spec.InternalUsers
+		// If not set, create empty map
+		if inUsers == nil {
+			inUsers = make(map[string]ElasticsearchUserSpec)
+		}
+
+		// "elastic" user
+		if userSpec, exists := inUsers[string(ElasticsearchInternalUserElastic)]; !exists {
+			inUsers[string(ElasticsearchInternalUserElastic)] = ElasticsearchUserSpec{
+				Reserved:     true,
```
Comment (kamolhasan) [inline apis/kubedb/v1alpha2/elasticsearch_helpers.go:608]: No need to use `reserved` here.

File: apis/kubedb/v1alpha2/elasticsearch_helpers.go [hunk 4]
```diff
@@ -595,6 +595,70 @@ func (e *Elasticsearch) setDefaultInternalUsersAndRoleMappings(esVersion *catalo
...
+			Reserved:     true,
+			BackendRoles: []string{"remote_monitoring_collector", "remote_monitoring_agent"},
+		})
+		setMissingElasticsearchUser(inUsers, string(ElasticsearchInternalUserLogstashSystem), ElasticsearchUserSpec{
+			Reserved:     true,
+			BackendRoles: []string{"logstash_system"},
+		})
+
+		// Set missing user secret names
+		for username, userSpec := range inUsers {
```
Comment (kamolhasan) [inline apis/kubedb/v1alpha2/elasticsearch_helpers.go:641]: Reuse the existing loop

File: apis/kubedb/v1alpha2/elasticsearch_types.go [hunk 5]
```diff
@@ -240,6 +245,23 @@ type ElasticsearchUserSpec struct {
 	// +optional
 	Hash string `json:"-" yaml:"hash,omitempty"`
 
+	// Specifies The full name of the user
+	// Only applicable for xpack-* versions
+	FullName string `json:"full_name,omitempty" yaml:"-"`
+
+	// Specifies Arbitrary metadata that you want to associate with the user
+	// Only applicable for xpack-* versions
```
Comment (kamolhasan) [inline apis/kubedb/v1alpha2/elasticsearch_types.go:253]: with auth plugin x-pack.

**Review by kamolhasan — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/constants.go (hunk 1 above)
Comment (raihankhan) [reply apis/kubedb/v1alpha2/constants.go:362]: renamed and moved

File: apis/kubedb/v1alpha2/elasticsearch_helpers.go (hunk 2 above)
Comment (raihankhan) [reply apis/kubedb/v1alpha2/elasticsearch_helpers.go:598]: Added

File: apis/kubedb/v1alpha2/elasticsearch_helpers.go (hunk 3 above)
Comment (raihankhan) [reply apis/kubedb/v1alpha2/elasticsearch_helpers.go:608]: Removed all occurances

File: apis/kubedb/v1alpha2/elasticsearch_helpers.go (hunk 4 above)
Comment (raihankhan) [reply apis/kubedb/v1alpha2/elasticsearch_helpers.go:641]: Reformatted

File: apis/kubedb/v1alpha2/elasticsearch_types.go (hunk 5 above)
Comment (raihankhan) [reply apis/kubedb/v1alpha2/elasticsearch_types.go:253]: fixed comments

File: apis/kubedb/v1alpha2/elasticsearch_types.go [hunk 6]
```diff
@@ -206,40 +206,65 @@ type ElasticsearchNode struct {
 	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
 }
 
-// +kubebuilder:validation:Enum=ca;transport;http;admin;archiver;metrics-exporter
+// +kubebuilder:validation:Enum=ca;transport;http;admin;client;metrics-exporter
```
Comment (kamolhasan) [inline apis/kubedb/v1alpha2/elasticsearch_types.go:209]: Keep, `archiver` alias and enum. 
Cleanup it from set-default

**Review by kamolhasan — APPROVED** [review]
LGTM!

**Pull request merged**